### DISCUSSION TITLE
Best way to handle IO bound and CPU bound task in a single route

### DISCUSSION THREAD

**Discussion by hari10599** [fastapi/fastapi#5969] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI
import requests
import httpx
from starlette.concurrency import run_in_threadpool

app = FastAPI()

def classify(content):
    '''
    Blocking code that does ML inference say image classification
    '''

@app.get("/sync")
def room_classify(url: str):
    response = requests.get(url)
    classify(response.content)

@app.get("/async")
async def room_classify(url: str):
    async with httpx.AsyncClient() as client:
        response = await client.get(url)
    classify(response.content)

@app.get("/async-tp")
async def room_classify(url: str):
    async with httpx.AsyncClient() as client:
        response = await client.get(url)
    await run_in_threadpool(classify, response.content)
```

### Description

Which is the best option to host an image classifier in Fastapi?

### Operating System

macOS

### FastAPI Version

0.89.1

### Python Version

Python 3.8.13

**Comment by mr-st0rm** [comment]
it depends

u can run it in new thread or process

**Comment by DurandA** [comment, marked as answer]
You would most likely want to run your CPU bound task (`classify`) in a new process and not a thread due to the infamous Python GIL.

This can be done using `run_in_executor` that returns a `Future`. Note that `run_in_executor` uses a `ThreadPoolExecutor` by default so you have to specify that you want a `ProcessPoolExecutor`:
```python
with ProcessPoolExecutor() as executor:
        result = await loop.run_in_executor(executor, classify, content)
```

**Comment by yinziyan1206** [comment]
you can use other process to manage cpu bound task (just like `Celery`, `ProcessPool` or etc) and use fastapi to wait the callback.
